# 第2课：NumPy入门

> 授课教师： [Yuki Oyama](mailto:y.oyama@lrcs.ac), [Prprnya](mailto:nya@prpr.zip)
>
> 克里斯蒂安·弗雷德里希·魏希曼化学系, 拉斯托利亚皇家理学院

本材料采用<a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh-hans">知识共享 署名-非商业性使用-相同方式共享 4.0</a> 许可协议授权<img src="https://mirrors.creativecommons.org/presskit/icons/cc.svg" alt="" style="max-width: 1em;max-height:1em;margin-left: .2em;"><img src="https://mirrors.creativecommons.org/presskit/icons/by.svg" alt="" style="max-width: 1em;max-height:1em;margin-left: .2em;"><img src="https://mirrors.creativecommons.org/presskit/icons/nc.svg" alt="" style="max-width: 1em;max-height:1em;margin-left: .2em;"><img src="https://mirrors.creativecommons.org/presskit/icons/sa.svg" alt="" style="max-width: 1em;max-height:1em;margin-left: .2em;">

In this lesson, we will introduce some basic functionalities of the `numpy` library, which is a powerful library for numerical computing in Python. It provides support for arrays, matrices, and a wide range of mathematical functions. In this lesson, we will cover some of the basic features of NumPy.


## Importing Libraries

In the first lesson, we introduced some basic concepts of variables and operations in Python, but in practice, we often need to use additional **libraries** to perform more complex tasks. Libraries are collections of pre-written code that provide additional functionality to Python. In this lesson, we will use the popular library: `numpy`.

To use a library, we need to import it first. We can do this using the `import` statement. For example, to import the `numpy` library, we can use the following code:

```python
import numpy
```

In [1]:
import numpy

After importing `numpy`, we can use its functions and methods to perform various operations on arrays and matrices. For example, we can check the value of $\pi$ (`pi`) using `numpy`:

```python
pi
```

In [2]:
# pi

Oh, no... What's going on? You should see a `NameError` after running the cell. This _error_ indicates that `pi` is not defined. In fact, `pi` is a part of the `numpy` library, and we need to specify that we are using it from that library, which can be done by prefixing it with `numpy.`:

```python
numpy.pi
```

In [3]:
numpy.pi

3.141592653589793

However, typing `numpy.` every time can be tedious. To make it easier, we can import the library with an _alias_ using the `as` keyword. This allows us to use a shorter name to refer to the library. For example, we can import `numpy` as `np`:

```python
import numpy as np
```

In [4]:
import numpy as np

You can even use `nyanya` as an alias if you want—

```python
import numpy as nyanya
nyanya.pi
```

In [5]:
import numpy as nyanya
nyanya.pi

3.141592653589793

For the sake of our (and more possibly, other people's!) mental health, let's stick to `np` as the alias for `numpy`, which is the most commonly used alias in the scientific Python community. Now, you can check the value of π again:

```python
import numpy as np
np.pi
```

In [6]:
import numpy as np
np.pi

3.141592653589793

Did you see the advantage of using aliases? It saves a lot of typing!

## Common Math Functions

NumPy provides a wide range of mathematical functions:

| Function     | Description                               |
|--------------|-------------------------------------------|
| `np.sin(x)`  | Sine of `x` ($\sin x$, $x$ in radians)    |
| `np.cos(x)`  | Cosine of `x` ($\cos x$, $x$ in radians)  |
| `np.tan(x)`  | Tangent of `x` ($\tan x$, $x$ in radians) |
| `np.exp(x)`  | Exponential of `x` ($e^x$)                |
| `np.log(x)`  | Natural logarithm of `x` ($\ln x$)        |
| `np.sqrt(x)` | Square root of `x` ($\sqrt{x}$)           |

For reference to more functions, see the [NumPy documentation](https://numpy.org/doc/stable/reference/routines.math.html).

Let's try this:

```python
np.sqrt(2)
```

In [7]:
np.sqrt(2)

1.4142135623730951

Also for this:

```python
np.exp(1)
```

In [8]:
np.exp(1)

2.718281828459045

...and then:

```python
np.cos(np.pi/2)
```

In [9]:
np.cos(np.pi/2)

6.123233995736766e-17

You will notice that the last output is not exactly zero. This is due to the limitations of floating-point arithmetic in computers, which can lead to small numerical errors. In practice, we often consider values very close to zero (e.g., within a small tolerance) as effectively zero.

## Arrays


NumPy provides a powerful data structure called an **array**, which is a collection of elements of the _same type_. Arrays can be one-dimensional (like a list), two-dimensional (like a matrix), or even multidimensional. Let's see an example of creating a one-dimensional array:

$$
\begin{bmatrix}1 & 2 & 3 & 4 & 5\end{bmatrix}
$$

```python
arr = np.array([1, 2, 3, 4, 5])
arr
```

In [10]:
arr = np.array([1, 2, 3, 4, 5])
arr

array([1, 2, 3, 4, 5])

In the code above, we created a one-dimensional array called `arr` using the `np.array()` function. The elements of the array are enclosed in square brackets and separated by commas. We can check the type of the variable `arr` using the `type()` function:
```python
type(arr)
```

In [11]:
type(arr)

numpy.ndarray

The output shows that `arr` is of type `numpy.ndarray`, which stands for "n-dimensional array". This is rather a different type than the first lesson (`int`, `float`, etc.), but we don't need to care the details.

We can also check the **size** of the array using the `size` attribute of the array, in which you need to add `.size` after the array name:

```python
arr.size
```

In [12]:
arr.size

5

Since arrays can only contain elements of the same type, if we try to create an array with mixed types, NumPy will automatically convert all elements to a common type. For example:

```python
arr_mixed = np.array([1, 2.5, 3, 4.0, 5])
arr_mixed
```

In [13]:
arr_mixed = np.array([1, 2.5, 3, 4.0, 5])
arr_mixed

array([1. , 2.5, 3. , 4. , 5. ])

```python
arr_mixed2 = np.array([1, 'two', 3, 4.0, 5])
arr_mixed2
```

In [14]:
arr_mixed2 = np.array([1, 'two', 3, 4.0, 5])
arr_mixed2

array(['1', 'two', '3', '4.0', '5'], dtype='<U32')

### Basic Operations

We can do some basic operations between arrays and numbers, such as addition, subtraction, multiplication, and division. For example:

$$
\begin{bmatrix}1+1 & 2+1 & 3+1 & 4+1 & 5+1\end{bmatrix} = \begin{bmatrix}2 & 3 & 4 & 5 & 6\end{bmatrix}
$$

```python
arr + 1
```

In [15]:
arr + 1

array([2, 3, 4, 5, 6])

$$
\begin{bmatrix}1 \times 2 & 2 \times 2 & 3 \times 2 & 4 \times 2 & 5 \times 2\end{bmatrix} = \begin{bmatrix}2 & 4 & 6 & 8 & 10\end{bmatrix}
$$

```python
arr * 2
```

In [16]:
arr * 2

array([ 2,  4,  6,  8, 10])

$$
\begin{bmatrix}1/3 & 2/3 & 3/3 & 4/3 & 5/3\end{bmatrix} = \begin{bmatrix}0.\dot{3} & 0.\dot{6} & 1 & 1.\dot{3} & 1.\dot{6}\end{bmatrix}
$$

```python
arr / 3
```

In [17]:
arr / 3

array([0.33333333, 0.66666667, 1.        , 1.33333333, 1.66666667])

We can also perform element-wise operations between two arrays of the same size. For example:

$$
\begin{bmatrix}10 & 20 & 30 & 40 & 50\end{bmatrix} + \begin{bmatrix}1 & 2 & 3 & 4 & 5\end{bmatrix} = \begin{bmatrix}11 & 22 & 33 & 44 & 55\end{bmatrix}
$$

```python
arr1 = np.array([10, 20, 30, 40, 50])
arr2 = np.array([1, 2, 3, 4, 5])
arr1 + arr2
```

In [18]:
arr1 = np.array([10, 20, 30, 40, 50])
arr2 = np.array([1, 2, 3, 4, 5])
arr1 + arr2

array([11, 22, 33, 44, 55])

$$
\begin{bmatrix}10 & 20 & 30 & 40 & 50\end{bmatrix} - \begin{bmatrix}1 & 2 & 3 & 4 & 5\end{bmatrix} = \begin{bmatrix}9 & 18 & 27 & 36 & 45\end{bmatrix}
$$

```python
arr1 - arr2
```

In [19]:
arr1 - arr2

array([ 9, 18, 27, 36, 45])

$$
\begin{bmatrix}10 \times 1 & 20 \times 2 & 30 \times 3 & 40 \times 4 & 50 \times 5\end{bmatrix} = \begin{bmatrix}10 & 40 & 90 & 160 & 250\end{bmatrix}
$$

```python
arr1 * arr2
```

In [20]:
arr1 * arr2

array([ 10,  40,  90, 160, 250])

$$
\begin{bmatrix}10/1 & 20/2 & 30/3 & 40/4 & 50/5\end{bmatrix} = \begin{bmatrix}10 & 10 & 10 & 10 & 10\end{bmatrix}
$$

```python
arr1 / arr2
```

In [21]:
arr1 / arr2

array([10., 10., 10., 10., 10.])

We can also use some common mathematical functions on arrays. For example:

$$
\begin{bmatrix}\sin 0 & \sin \dfrac{\pi}{2} & \sin \pi\end{bmatrix} = \begin{bmatrix}0 & 1 & 0\end{bmatrix}
$$

```python
arr3 = np.array([0, np.pi/2, np.pi])
np.sin(arr3)
```

In [22]:
arr3 = np.array([0, np.pi/2, np.pi])
np.sin(arr3)

array([0.0000000e+00, 1.0000000e+00, 1.2246468e-16])

### Indexing and Slicing

Sometimes we want to pick a single element from an array. We can do this using **indexing**. <u>**In Python, indexing starts from `0`.**</u> Indices must be integers. For example, to access the first element of `arr`, we can use:

```python
arr[0]
```

In [23]:
arr[0]

1

```python
arr[1]
```

In [24]:
arr[1]

2

```python
arr[4]
```

In [25]:
arr[4]

5

```python
arr[-1]
```

In [26]:
arr[-1]

5

See what's going on for this `arr[-1]`? The index `-1` refers to the last element of the array. Similarly, `-2` refers to the second-to-last element, and so on.

What if we try to access an index out of range?

```python
arr[5]
```

In [27]:
# arr[5]

Like the case that we want to access `pi` without importing `numpy`, we again see an _error_ here, but this is a different type of error, an `IndexError`, which indicates that the index is out of range. In this case, the valid indices for `arr` are `0` to `4` (or `-1` to `-5`), so trying to access index `5` results in an error. _We will discuss error handling later._

If we want to access multiple elements from an array, we can use **slicing**. Slicing allows us to extract a portion of the array by specifying a range of indices. The syntax for slicing is `array[start:stop]`, where `start` is the index of the first element to <u>**include**</u>, and `stop` is the index of the first element to <u>**exclude**</u>. Since `start` and `stop` are indices, they must be integers. For example:

```python
arr[1:4]
```

In [28]:
arr[1:4]

array([2, 3, 4])

```python
arr[0:3]
```

In [29]:
arr[0:3]

array([1, 2, 3])

```python
arr[2:5]
```

In [30]:
arr[2:5]

array([3, 4, 5])

If we set `start` equal to `stop`, we get an empty array:

```python
arr[3:3]
```

In [31]:
arr[3:3]

array([], dtype=int32)

Obviously, if `start` is greater than `stop`, we also get an empty array:

```python
arr[4:2]
```

In [32]:
arr[4:2]

array([], dtype=int32)

If we omit the `start` index, it defaults to `0`. If we omit the `stop` index, it defaults to the length of the array. For example:

```python
arr[:3]
```

In [33]:
arr[:3]

array([1, 2, 3])

```python
arr[2:]
```

In [34]:
arr[2:]

array([3, 4, 5])

Definitely, if we omit both `start` and `stop`, we get the entire array:

```python
arr[:]
```

In [35]:
arr[:]

array([1, 2, 3, 4, 5])

In addition to `start` and `stop`, we can also specify a `step` value using the syntax `array[start:stop:step]`. The `step` value determines the interval between elements to include in the slice. To avoid confusion, `step` must be a non-zero integer. If `step` is omitted, it defaults to `1`. For example:

```python
arrlong = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
arrlong[2:7:2]
```

In [36]:
arrlong = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
arrlong[2:7:2]

array([3, 5, 7])

```python
arrlong[::3]
```

In [37]:
arrlong[::3]

array([ 1,  4,  7, 10])

```python
arrlong[2:5:1]
```

In [38]:
arrlong[2:5:1]

array([3, 4, 5])

For negative `step`, the slice is taken in _reverse_ order. For example:

```python
arrlong[::-1]
```

In [39]:
arrlong[::-1]

array([10,  9,  8,  7,  6,  5,  4,  3,  2,  1])

If we use a negative `step`, we need to ensure that the `start` index is greater than the `stop` index; otherwise, we will get an empty array. For example:

```python
arrlong[7:2:-2]
```

In [40]:
arrlong[7:2:-2]

array([8, 6, 4])

```python
arrlong[2:7:-2]
```

In [41]:
arrlong[2:7:-2]

array([], dtype=int32)

<span style="color:green">**Exercise**:</span> The code below generates an array of integers from 1 to 20.

```python
arr_ex = np.arange(1, 21)
```

Write code to implement the following tasks:

  -  Extract the elements at odd indices (i.e., indices 1, 3, 5, ...) from this array using slicing.
  -  Subtract 1 from each of these extracted elements.
  -  Multiply each of the resulting elements by $\pi/6$.
  -  Calculate the cosine of each of these modified elements using `np.cos()`.

Print the results of each step to verify your work.

## 2D Arrays

NumPy also supports multidimensional arrays. A two-dimensional array can be thought of as a [matrix](https://en.wikipedia.org/wiki/Matrix_(mathematics)), with rows and columns. We can create a two-dimensional array using the `np.array()` function, by passing a list of lists. For example:

$$
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{bmatrix}
$$

```python
matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
matrix
```

In [42]:
matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
matrix

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

Like one-dimensional arrays, we can check the type of the two-dimensional array:

```python
type(matrix)
```

In [43]:
type(matrix)

numpy.ndarray

We notice that the type is still `numpy.ndarray`, which means that NumPy does not differentiate between one-dimensional and multidimensional arrays in terms of type. We can also check the size of the two-dimensional array, which is the total number of elements in the array:

```python
matrix.size
```

In [44]:
matrix.size

9

For 2D arrays, we can also check the **shape** of the array using the `shape` attribute. The shape gives the number of rows and columns in the array in the form `(rows, columns)`:

```python
matrix.shape
```

In [45]:
matrix.shape

(3, 3)

The same operations we discussed for one-dimensional arrays can also be applied to two-dimensional arrays. For example, we can perform element-wise addition, subtraction, multiplication, and division between two matrices of the same shape. We can also use mathematical functions on each element of the matrix.

$$
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{bmatrix}
+
\begin{bmatrix}
9 & 8 & 7 \\
6 & 5 & 4 \\
3 & 2 & 1
\end{bmatrix}
=
\begin{bmatrix}
10 & 10 & 10 \\
10 & 10 & 10 \\
10 & 10 & 10
\end{bmatrix}
$$

```python
matrix2 = np.array([[9, 8, 7], [6, 5, 4], [3, 2, 1]])
matrix + matrix2
```

In [46]:
matrix2 = np.array([[9, 8, 7], [6, 5, 4], [3, 2, 1]])
matrix + matrix2

array([[10, 10, 10],
       [10, 10, 10],
       [10, 10, 10]])

$$
\begin{bmatrix}
\sqrt{1} & \sqrt{2} & \sqrt{3} \\
\sqrt{4} & \sqrt{5} & \sqrt{6} \\
\sqrt{7} & \sqrt{8} & \sqrt{9}
\end{bmatrix}
=
\begin{bmatrix}
1 & \sqrt{2} & \sqrt{3} \\
2 & \sqrt{5} & \sqrt{6} \\
\sqrt{7} & 2\sqrt{2} & 3
\end{bmatrix}
\approx
\begin{bmatrix}
1 & 1.414 & 1.732 \\
2 & 2.236 & 2.449 \\
2.646 & 2.828 & 3
\end{bmatrix}
$$

```python
np.sqrt(matrix)
```

In [47]:
np.sqrt(matrix)

array([[1.        , 1.41421356, 1.73205081],
       [2.        , 2.23606798, 2.44948974],
       [2.64575131, 2.82842712, 3.        ]])

Indexing and slicing also work similarly for two-dimensional arrays. We can access individual elements, rows, or columns using indexing and slicing. Use the following matrix as an example:

```python
a = np.array([[ 0,  1,  2,  3,  4,  5],
              [10, 11, 12, 13, 14, 15],
              [20, 21, 22, 23, 24, 25],
              [30, 31, 32, 33, 34, 35],
              [40, 41, 42, 43, 44, 45],
              [50, 51, 52, 53, 54, 55]])
a
```

In [48]:
a = np.array([[ 0,  1,  2,  3,  4,  5],
              [10, 11, 12, 13, 14, 15],
              [20, 21, 22, 23, 24, 25],
              [30, 31, 32, 33, 34, 35],
              [40, 41, 42, 43, 44, 45],
              [50, 51, 52, 53, 54, 55]])
a

array([[ 0,  1,  2,  3,  4,  5],
       [10, 11, 12, 13, 14, 15],
       [20, 21, 22, 23, 24, 25],
       [30, 31, 32, 33, 34, 35],
       [40, 41, 42, 43, 44, 45],
       [50, 51, 52, 53, 54, 55]])

To access a single row, we can use a single index. For example, to access the second row (which is `[10, 11, 12, 13, 14, 15]`), we can use:

```python
a[1]
```

In [49]:
a[1]

array([10, 11, 12, 13, 14, 15])

To access a single element, we should use two indices: <u>the first index for the row</u> and <u>the second index for the column</u>. For example, to access the element in the second row and third column (which is `12`), we can use:

```python
a[1][2]
```

In [50]:
a[1][2]

12

A shortcut of doing this picking is:

```python
a[1, 2]
```

This only applies to NumPy arrays, not Python lists.

In [51]:
a[1, 2]

12

However, accessing a single column is a bit tricky (for Python lists). Luckily, NumPy provides a convenient way to do this using slicing. Expanding the usage of the indexing shortcut above, we can replace any of the indices by a slice (`start:stop:step`).

Let's see some examples:

<img src="./assets/numpy_indexing.png" alt="2D array" width="50%" style="display=block; margin:auto"/>

> Image source: [NumPy documentation](https://lectures.scientific-python.org/intro/numpy/array_object.html#indexing-and-slicing)

  -  The <span style="color:red">red</span> one is `a[0, 3:5]`, which picks <u>the first row</u> on <u>the 4th to 5th columns</u> (excluding the 6th column).
  -  The <span style="color:green">green</span> one is `a[4:, 4:]`, which picks <u>the 5th to the last rows</u> on <u>the 5th to the last columns</u>.
  -  The <span style="color:blue">blue</span> one is `a[:, 2]`, which picks <u>all the rows</u> on <u>the 2nd column</u>.
  -  The <span style="color:purple">purple</span> one is `a[2::2, ::2]`, which picks <u>the 3rd to the last rows with a step of 2</u> on <u>all the columns with a step of 2</u>.

To verify these, you can run the following codes:

```python
print(a[0, 3:5])    # Red
print(a[4:, 4:])    # Green
print(a[:, 2])      # Blue
print(a[2::2, ::2]) # Purple
```

In [52]:
print(a[0, 3:5])    # Red
print(a[4:, 4:])    # Green
print(a[:, 2])      # Blue
print(a[2::2, ::2]) # Purple

[3 4]
[[44 45]
 [54 55]]
[ 2 12 22 32 42 52]
[[20 22 24]
 [40 42 44]]


<span style="color:green">**Exercise**:</span> Using matrix slicing, extract the submatrix from the matrix `a` that includes the elements from the 2nd to the 4th rows and the 3rd to the 5th columns (inclusive of the start index and exclusive of the stop index, basically, a 3*3 submatrix centered around the element `23`). Print the resulting submatrix.

## Two Useful Functions for Arrays

Here are two useful functions to create arrays with evenly spaced values:

  -  `np.arange(start, stop, step)`: Creates an array with evenly spaced values within a specified range.
  -  `np.linspace(start, stop, num)`: Creates an array with a specified number of evenly spaced values between two endpoints.

For example, to create an array of _integers_ from 0 to 9, we can use `np.arange()`:

```python
np.arange(0, 10, 1)
```

In [53]:
np.arange(0, 10, 1)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

To create an array of _odd numbers_ from 5 to 15, we can use:

```python
np.arange(5, 16, 2)
```

In [54]:
np.arange(5, 16, 2)

array([ 5,  7,  9, 11, 13, 15])

For making an array from 0 to 1 with 5 evenly spaced values, we can use `np.linspace()`:

```python
np.linspace(0, 1, 5)
```

In [55]:
np.linspace(0, 1, 5)

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

Finally, let's create an array of 4 evenly spaced values from 0 to $\pi$:

```python
np.linspace(0, np.pi, 4)
```

In [56]:
np.linspace(0, np.pi, 4)

array([0.        , 1.04719755, 2.0943951 , 3.14159265])

## End-of-Lesson Problems

### Problem 1: Play with Arrays

Using only methods introduced in this lesson:

1. Create a 1D array `radians` of 9 evenly spaced values from 0 to π (inclusive) using `np.linspace`.

2. Compute two arrays:
   - `s = np.sin(radians)`
   - `c = np.cos(radians)`

3. Form `t = s*s + c*c`. Using slicing, extract every second element of `t` starting from index `1`.

4. Using the existing 2D array `matrix` (already defined in the notebook), do the following:
   - Report `matrix.size` and `matrix.shape`.
   - Extract the last column using slicing/indexing.
   - Extract the submatrix consisting of all rows and the first two columns.

### Problem 2: Quantum Harmonic Oscillator

Consider a one-dimensional [quantum harmonic oscillator](https://en.wikipedia.org/wiki/Quantum_harmonic_oscillator), but with a twist: the potential energy is added with a [Dirac delta function](https://en.wikipedia.org/wiki/Dirac_delta_function) at the origin. The potential is given by:

$$V(x) = \frac{1}{2} m \omega^2 x^2 + \alpha \delta(x)$$

where $m$ is the mass of the particle, $\omega$ is the angular frequency, and $\alpha$ is a constant representing the strength of the delta potential. For simplicity, we set $m = 1$, $\omega = 1$, and $\alpha = 1$. That is,

$$V(x) = \frac{1}{2} x^2 + \delta(x)$$

The wavefunctions of the unperturbed harmonic oscillator are given by:

$$\psi_n(x) = \left(\frac{m \omega}{\pi \hbar}\right)^{1/4} \frac{1}{\sqrt{2^n n!}} H_n\left(\sqrt{\frac{m \omega}{\hbar}} x\right) e^{-\frac{m \omega x^2}{2 \hbar}}$$

where $H_n$ are the [Hermite polynomials](https://zh.wikipedia.org/zh-cn/%E5%9F%83%E5%B0%94%E7%B1%B3%E7%89%B9%E5%A4%9A%E9%A1%B9%E5%BC%8F). For simplicity, we set $\hbar = 1$. Substituting the values of $m$, $\omega$, and $\hbar$, we have:

$$\psi_n(x) = \left(\frac{1}{\pi}\right)^{1/4} \frac{1}{\sqrt{2^n n!}} H_n(x) e^{-\frac{x^2}{2}}$$

Using NumPy, perform the following tasks:

1. Create a 1D array `x` of 200 evenly spaced values from -5 to 5 using `np.linspace`.

2. The code below defines a function `harmonic_psi(n, x)` that computes the wavefunction $\psi_n(x)$ for a given quantum number `n` and position array `x`. Test the value of $\psi_0$ at $x = 0$ and $x = 1$.

    ```python
    from scipy.special import eval_hermite, factorial

    def harmonic_psi(n, x):
        return eval_hermite(n, x) * np.exp(-x ** 2 / 2) / np.sqrt(2 ** n * factorial(n)) / (np.pi) ** 0.25
    ```

3. Compute the values of $\psi_0$ from $x=-5$ to $x=5$ using the array `x` created before, and store the result in `psi_0`. Similarly, compute `psi_1` to `psi_5` for the first to fifth excited states. Create a 2D array `psis` where each row corresponds to a different quantum state (from $n=0$ to $n=5$) and each column corresponds to a position in `x`.

4. Now it's time to bring our delta potential into play. The presence of the delta function at the origin modifies the wavefunctions, but it only affects the even states ($n=0, 2, 4, ...$). It's a little bit more complicated to derive the even state solutions, but we can still keep our odd states. Create a new array `psis_odd` by slicing `psis` to keep only the odd states ($n=1, 3, 5$).

5. Run the code below, and you should be able to visualize the wavefunctions of the odd states. _We will learn how to make plots like this later._

    ```python
    from matplotlib import pyplot as plt

    plt.figure(figsize=(8, 6))

    for n in range(psis_odd.shape[1]):
        plt.plot(x, psis_odd[:, n], label=f'n={n}')

    plt.grid(linestyle='--')

    plt.legend()
    plt.show()
    ```

## Acknowledgement

This lesson draws on ideas from the following sources:

- [NumPy Official Website](https://numpy.org)
- [Scientific Python Lectures](https://lectures.scientific-python.org/)
- Charles J. Weiss's [Scientific Computing for Chemists with Python](https://weisscharlesj.github.io/SciCompforChemists/notebooks/introduction/intro.html)
- [An Introduction to Python for Chemistry](https://pythoninchemistry.org/intro_python_chemists/intro.html)
- GenAI for making paragraphs and codes(・ω< )★
- And so many resources on Reddit, StackExchange, etc.!